src/combinatorics/simple_graph/subgraph.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2021 Hunter Monroe. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hunter Monroe, Kyle Miller, Alena Gusakov
+-/
+
+import combinatorics.simple_graph.basic
+
+/-!
+# Subgraphs of a simple graph
+
+This file formalizes a subgraph of a simple graph as a structure and proves a few simple lemmas.
+-/
+
+universe u
+variables {V : Type u} (G : simple_graph V)
+open simple_graph
+
+/--
+A subgraph of a graph is a subset of vertices and a subset of edges such that each endpoint of
+an edge is contained in the vertex set. Subgraphs implement the `simple_graph` class.
+-/
+@[ext]
+structure subgraph :=
+(verts : set V)
+(adj' : V → V → Prop)
+(adj_sub : ∀ ⦃v w : V⦄, adj' v w → G.adj v w)
+(edge_vert : ∀ ⦃v w : V⦄, adj' v w → v ∈ verts)
+(sym' : symmetric adj')
+
+/--
+The Prop that states that `G'` is isomorphic to a subgraph of `G`.
+-/
+def is_subgraph (G' : simple_graph V) : Prop := ∀ ⦃v w : V⦄, G'.adj v w → G.adj v w
+
+namespace subgraph
+
+variable {G}
+
+/--
+The edges of `G'` consist of a subset of edges of `G`.
+-/
+def edge_set' (G' : subgraph G) : set (sym2 V) := sym2.from_rel G'.sym'
+
+@[simp]
+lemma mem_edge_set' {G' : subgraph G} {v w : V} : ⟦(v, w)⟧ ∈ edge_set' G' ↔ G'.adj' v w :=
+by refl
+
+lemma edge_sub (G' : subgraph G) : G'.edge_set' ⊆ G.edge_set :=
+begin
+  intro e,
+  refine quotient.rec_on_subsingleton e (λ e, _),
+  cases e with v w,
+  simp only [mem_edge_set', mem_edge_set],
+  apply G'.adj_sub,
+end
+
+lemma has_verts (G' : subgraph G) : ∀ ⦃e : sym2 V⦄ ⦃v : V⦄, e ∈ G'.edge_set' → v ∈ e →
```

Comment:
Move `e` (at least) to the left-hand side of the colon and remove the `intro e` from the proof.

Reply:
Could you provide a specific suggestion of code--I tried this and got an error with quotient.rec_on_subsingleton

Reply:
```lean
lemma has_verts (G' : subgraph G) {e : sym2 V} {v : V} (he : e ∈ G'.edge_set') (hv : v ∈ e) :
  v ∈ G'.verts :=
begin
  refine quotient.rec_on_subsingleton e (λ e v he hv, _) v he hv,
  cases e,
  rw mem_edge_set' at he,
  cases sym2.mem_iff.mp hv; subst v,
  { exact G'.edge_vert he },
  { exact G'.edge_vert (G'.sym' he) },
end
```